### PR TITLE
cleanup vram

### LINKED ISSUE #3638: activation_offloading: true causes linear VRAM leak with Gemma 4 26B-A4B (LoRA + QLoRA, with torch_compile + sample_packing)
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

GPU VRAM usage should remain approximately constant across training steps after warmup. 

### Current behaviour

When fine-tuning `google/gemma-4-26B-A4B` with `activation_offloading: true`, GPU VRAM grows linearly throughout training until OOM. The leak reproduces with both LoRA and QLoRA adapters. 

This is **not** the per-checkpoint spike pattern reported in #3596 — no checkpoint saves or evaluation are configured in the runs below. 

The yaml is essentially https://github.com/axolotl-ai-cloud/axolotl/blob/main/examples/gemma4/26b-a4b-moe-qlora.yaml 
I am using cu130 because of B300 see #3630 

### Steps to reproduce

Six near-identical configs were run side-by-side, differing only in adapter type (LoRA vs QLoRA) and one toggled feature:

| Run | Adapter | `activation_offloading` | `use_scattermoe` | Result |
|---|---|---|---|---|
| `qlora-control` | qlora | true | true | **leaks** linearly (24 → 260 GiB over 293 steps) |
| `qlora-no-actoffload` | qlora | **false** | true | **flat at 24 GiB** |
| `qlora-no-scattermoe` | qlora | true | **false** | flat at 100 GiB |
| `lora-control` | lora | true | true | **leaks** linearly (120 → 260 GiB over 224 steps) |
| `lora-no-actoffload` | lora | **false** | true | **flat at 60 GiB** |
| `lora-no-scattermoe` | lora | true | **false** | **leaks** linearly (120 → 260 GiB over 149 steps) |

`activation_offloading: false` is the only intervention that stops the leak in both LoRA and QLoRA. Disabling ScatterMoE only stops the leak in the QLoRA case.

See attached TensorBoard screenshots showing `train/memory/max_allocated`, `train/memory/device_reserved`, `train/memory/max_active`, and `train/loss` for both matrices.

QLora
Image 
Lora
Image

Dockerfile (cu130 was needed for  NVIDIA B300 SXM6 AC see issue #3630 )

```
FROM axolotlai/axolotl-uv:main-py3.11-cu130-2.9.1

RUN uv pip install --python /workspace/axolotl-venv/bin/python \
    "cut-cross-entropy[transformers] @ git+https://github.com/axolotl-ai-cloud/ml-cross-entropy.git@fec1a88"
```

Docker run
```
# launch train
docker run --rm --gpus "$GPU_SPEC" --ipc=host \
  --shm-size 10g \
  -e TRITON_PTXAS_PATH=/usr/local/cuda/bin/ptxas \
  -e HF_TOKEN="$HF_TOKEN" \
  -v "$HOME/projects/project-training:/workspace/project" \
  -v "$HOME/.cache/huggingface:/root/.cache/huggingface" \
  -w /workspace/project \
  project-axolotl \
  axolotl train "$CONFIG"
```

yamls
[bug_lora-noactoffload.yaml](https://github.com/user-attachments/files/27277714/bug_lora-noactoffload.yaml)
[bug_lora-noscattermoe.yaml](https://github.com/user-attachments/files/27277710/bug_lora-noscattermoe.yaml)
[bug_lora.yaml](https://github.com/user-attachments/files/27277711/bug_lora.yaml)
[bug_qlora-noactoffload.yaml](https://github.com/user-attachments/files/27277713/bug_qlora-noactoffload.yaml)
[bug_qlora-noscattermoe.yaml](https://github.com/user-attachments/files/27277709/bug_qlora-noscattermoe.yaml)
[bug_qlora.yaml](https://github.com/user-attachments/files/27277712/bug_qlora.yaml)

### Config yaml

```yaml
# BUG REPRO: Gemma 4 26B-A4B MoE QLoRA — CONTROL (expected: VRAM leak)

base_model: google/gemma-4-26B-A4B

plugins:
  - axolotl.integrations.cut_cross_entropy.CutCrossEntropyPlugin
  - axolotl.integrations.kernels.KernelsPlugin
  - axolotl.integrations.liger.LigerPlugin
use_kernels: true
use_scattermoe: true
experts_implementation: scattermoe
torch_compile: true
liger_layer_norm: true
liger_rope: true
liger_rms_norm: true
liger_glu_activation: true
liger_rms_norm_gated: true
strict: false

chat_template: gemma4
datasets:
  - path: mlabonne/FineTome-100k
    type: chat_template
    split: train[:10%]
    field_messages: conversations
    message_property_mappings:
      role: from
      content: value

val_set_size: 0
output_dir: ./outputs/bug/qlora-control

sequence_len: 2048
sample_packing: true

load_in_4bit: true
quantize_moe_experts: true
adapter: qlora
lora_r: 16
lora_alpha: 32
lora_dropout: 0
lora_target_modules: 'model.language_model.layers.[\d]+.(_checkpoint_wrapped_module.)?(mlp|self_attn).(up|down|gate|q|k|v|o)_proj'
lora_target_parameters:
  - experts.gate_up_proj
  - experts.down_proj
lora_mlp_kernel: false
lora_qkv_kernel: false
lora_o_kernel: false

bnb_config_kwargs:
  bnb_4bit_use_double_quant: true

gradient_accumulation_steps: 4
micro_batch_size: 1
num_epochs: 1
optimizer: adamw_torch_8bit
lr_scheduler: cosine
learning_rate: 0.0002

bf16: auto
tf32: true

gradient_checkpointing: true
activation_offloading: true

sdp_attention: true

warmup_ratio: 0.1
evals_per_epoch: 0
saves_per_epoch: 0
save_strategy: "no"
weight_decay: 0.0
special_tokens:

logging_steps: 1
use_tensorboard: true
run_name: qlora-control
```

### Possible solution

_No response_

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.11

### axolotl branch-commit

main @ 6136ae6 (image: axolotlai/axolotl-uv:main-py3.11-cu130-2.9.1)

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

### PR DESCRIPTION
**Pull request by ved1beta**
# What does this PR do?

`OffloadActivations` keeps mutable state (`tracker`, `storage_to_tensor_id`, `fwd_stash`, `bwd_tensor_stash`, `bwd_ev_stash`, plus `is_first_*` booleans) across forward/backward. The reset is gated on `tracker` becoming empty during the unpack of the last saved tensor — but if any saved tensor never unpacks (its backward node never executes), the reset never fires. Each subsequent step adds more leaked entries, pinning GPU activation references → linear VRAM leak.

Hits reliably on **MoE + sample_packing + torch.compile** (e.g. Gemma‑4 26B‑A4B + ScatterMoE), where dynamic expert routing leaves saved tensors on subgraphs that don't contribute to the loss.

**Fix:** override `__enter__` to clear stale state before re-entering. By the time the next step enters the context, the previous backward has completed (Python is synchronous), so anything still in the dicts is leaked and safe to drop. 16 lines, no other logic touched.

Validated on the failing Gemma‑4 26B‑A4B QLoRA config (axolotl `examples/gemma4/26b-a4b-moe-qlora.yaml`): without the fix, ~60 saved tensors leak per micro‑step and OOM hits by step 2; with the fix, `tracker == 0` at the start of every step and VRAM stays flat across 200+ consecutive steps.

Fixes axolotl-ai-cloud/axolotl#3638

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case. — axolotl-ai-cloud/axolotl#3638
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## AI writing disclosure

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

---

> [!NOTE]
> **Medium Risk**
> Touches autograd `saved_tensors_hooks` lifecycle by resetting internal tracking on `__enter__`, which could affect activation offloading behavior if any tensors are unexpectedly needed across context boundaries; however the change is small and scoped to state cleanup.
> 
> **Overview**
> Prevents a step-to-step VRAM leak in `OffloadActivations` by overriding `__enter__` to **clear any leftover tracking/stash state** (`tracker`, `storage_to_tensor_id`, and stream stashes) and reset `is_first_*` flags before re-registering hooks.
> 
> This ensures activations that were saved in the forward but never unpacked in backward (e.g., conditional/MoE paths) don’t keep accumulating references and pin GPU memory across iterations.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit de13b4568f20ebbdc121df304fb070d430354cb0. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by butterwecksolutions** [comment]
Great find on the `__enter__` path! We independently ran into the QLoRA/BNB buffer leak and initially fixed it in `__exit__`, but your `__enter__` insight was the key — the class example shows `backward()` after `with` exits, so      
  `__exit__` can fire before backward completes. We've adopted your `__enter__` approach in #5730 and extended it with the BNB dequant buffer cleanup for the QLoRA path:
                                                                                                                                                                                                                                            
  | Capability | #5700 (merged) | #5738 | #5730 |                                                                                                                                                                                           
  |---|:-:|:-:|:-:|                                                                                                                                                                                                                         
  | CUDA stream sync + stash clear | ✅ | — | — |                                                                                                                                                                                           
  | tracker + storage_to_tensor_id clear | — | ✅ | ✅ |                                                                                                                                                                                    
  | stash clear (use_streams guard) | — | ✅ | ✅ |                                                                                                                                                                                         
  | is_first_forward/backward reset | — | ✅ | ✅ |                                                                                                                                                                                         
  | tensor_id reset | — | — | ✅ |                                                                                                                                                                                                          
  | BNB dequant buffer empty_cache | — | — | ✅ |                                                                                                                                                                                           
  | CUDA stream garbage leak | ✅ | — | — |                                                                                                                                                                                                 
  | MoE + compile leak path | — | ✅ | ✅ |                                                                                                                                                                                                 
  | QLoRA buffer leak path | — | — | ✅ |                                                                                                                                                                                                   
                                                                                                                                                                                                                                            
  Might make sense to consolidate review on #5730 since it covers the two residual leak paths from the `__enter__` you discovered.

**Comment by butterwecksolutions** [comment]
integrated and merged with https://github.com/huggingface/trl/pull/5730 please check

**Comment by albertvillanova** [comment]
Thanks for chasing this down, @ved1beta — and for the detailed Gemma-4 MoE repro. 

The exact fix here (overriding `__enter__` to clear `tracker`/`storage_to_tensor_id`/the stream stashes and reset the `is_first_*` flags) already landed in `main` via #5730, which was merged the same day you opened this. That PR is a strict superset: it also clears `_storage_key_by_tensor_id`, resets `tensor_id`, and flushes the bitsandbytes allocator cache for the QLoRA path. This branch actually descends from #5730, so it now defines `__enter__` twice — Python keeps the later (#5730) definition and silently shadows the one added here, making the diff a no-op. 

Cleanest path is to close this as already-fixed; the leak you reported should be gone on current `main`. Thanks again for the effort (also on the earlier #5737)!
